File: crates/dry_run_core/src/query/explain.rs
use serde::{Deserialize, Serialize};
use sqlx::PgPool;

use super::plan::{PlanNode, parse_plan_json};
use super::plan_warnings::detect_plan_warnings;
use crate::error::{Error, Result};
use crate::schema::AnnotatedSchema;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExplainResult {
    pub plan: PlanNode,
    pub total_cost: f64,
    pub estimated_rows: f64,
    pub warnings: Vec<PlanWarning>,
    pub execution: Option<ExecutionStats>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub raw_plan: Option<serde_json::Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pgmustard_tips: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlanWarning {
    pub severity: String,
    pub message: String,
    pub node_type: String,
    pub detail: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExecutionStats {
    pub execution_time_ms: f64,
    pub planning_time_ms: f64,
}

pub async fn explain_query(
    pool: &PgPool,
    sql: &str,
    analyze: bool,
    annotated: Option<&AnnotatedSchema<'_>>,
) -> Result<ExplainResult> {
    let explain_sql = if analyze {
        format!("EXPLAIN (ANALYZE, BUFFERS, FORMAT JSON) {sql}")
    } else {
        format!("EXPLAIN (FORMAT JSON) {sql}")
    };

    let json_str: String = if analyze {
        let mut tx = pool.begin().await?;

        let result: String = sqlx::query_scalar(&explain_sql)
            .fetch_one(&mut *tx)
            .await
            .map_err(|e| Error::Introspection(format!("EXPLAIN ANALYZE failed: {e}")))?;

        tx.rollback().await.ok();
        result
    } else {
        sqlx::query_scalar(&explain_sql)
            .fetch_one(pool)
            .await
            .map_err(|e| Error::Introspection(format!("EXPLAIN failed: {e}")))?
    };

    let plan_json: serde_json::Value = serde_json::from_str(&json_str)
        .map_err(|e| Error::Introspection(format!("failed to parse EXPLAIN JSON: {e}")))?;

    let plan_obj = plan_json
        .as_array()
        .and_then(|a| a.first())
        .ok_or_else(|| Error::Introspection("empty EXPLAIN result".into()))?;

    let plan_node_json = plan_obj
        .get("Plan")
        .ok_or_else(|| Error::Introspection("no Plan in EXPLAIN output".into()))?;

    let plan = parse_plan_json(plan_node_json)?;

    let total_cost = plan.total_cost;
    let estimated_rows = plan.plan_rows;

    let execution = if analyze {
        let exec_time = plan_obj
            .get("Execution Time")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        let plan_time = plan_obj
            .get("Planning Time")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.0);
        Some(ExecutionStats {
            execution_time_ms: exec_time,
            planning_time_ms: plan_time,
        })
    } else {
        None
    };

    let warnings = detect_plan_warnings(&plan, annotated);

    Ok(ExplainResult {
        plan,
        total_cost,
        estimated_rows,
        warnings,
        execution,
        raw_plan: Some(plan_json),
        pgmustard_tips: None,
    })
}

File: internal/query/explain.go
package query

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/boringsql/dryrun/internal/pgmustard"
	"github.com/boringsql/dryrun/internal/schema"
)

type (
	ExplainResult struct {
		Plan          PlanNode             `json:"plan"`
		TotalCost     float64              `json:"total_cost"`
		EstimatedRows float64              `json:"estimated_rows"`
		Warnings      []PlanWarning        `json:"warnings"`
		Execution     *ExecutionStats      `json:"execution,omitempty"`
		StatsInjected *schema.InjectResult `json:"stats_injected,omitempty"`
		GucsReplayed  []string             `json:"gucs_replayed,omitempty"`
		GucsSkipped   []SkippedGuc         `json:"gucs_skipped,omitempty"`
		PgMustardTips []pgmustard.Tip      `json:"pgmustard_tips,omitempty"`
		RawPlanJSON   json.RawMessage      `json:"-"`
	}

	PlanWarning struct {
		Severity string  `json:"severity"`
		Message  string  `json:"message"`
		NodeType string  `json:"node_type"`
		Detail   *string `json:"detail,omitempty"`
	}

	ExecutionStats struct {
		ExecutionTimeMs float64 `json:"execution_time_ms"`
		PlanningTimeMs  float64 `json:"planning_time_ms"`
	}
)

// runs EXPLAIN, optionally with ANALYZE in a rolled-back tx
func ExplainQuery(ctx context.Context, pool *pgxpool.Pool, sql string, analyze bool, snap *schema.SchemaSnapshot) (*ExplainResult, error) {
	var explainSQL string
	if analyze {
		explainSQL = fmt.Sprintf("EXPLAIN (ANALYZE, BUFFERS, FORMAT JSON) %s", sql)
	} else {
		explainSQL = fmt.Sprintf("EXPLAIN (FORMAT JSON) %s", sql)
	}

	replayGUCs := snap != nil && len(snap.GUCs) > 0

	var (
		jsonStr  string
		replayed []string
		skipped  []SkippedGuc
	)
	if analyze || replayGUCs {
		tx, err := pool.Begin(ctx)
		if err != nil {
			return nil, fmt.Errorf("begin transaction: %w", err)
		}
		defer tx.Rollback(ctx)

		if replayGUCs {
			replayed, skipped, err = applyPlannerGUCs(ctx, tx, snap.GUCs)
			if err != nil {
				return nil, fmt.Errorf("replay GUCs: %w", err)
			}
		}
		if err := tx.QueryRow(ctx, explainSQL).Scan(&jsonStr); err != nil {
			return nil, fmt.Errorf("EXPLAIN failed: %w", err)
		}
	} else {
		if err := pool.QueryRow(ctx, explainSQL).Scan(&jsonStr); err != nil {
			return nil, fmt.Errorf("EXPLAIN failed: %w", err)
		}
	}

	var planArray []json.RawMessage
	if err := json.Unmarshal([]byte(jsonStr), &planArray); err != nil {
		return nil, fmt.Errorf("failed to parse EXPLAIN JSON: %w", err)
	}
	if len(planArray) == 0 {
		return nil, fmt.Errorf("empty EXPLAIN result")
	}

	var planObj map[string]json.RawMessage
	if err := json.Unmarshal(planArray[0], &planObj); err != nil {
		return nil, fmt.Errorf("failed to parse EXPLAIN plan object: %w", err)
	}

	planRaw, ok := planObj["Plan"]
	if !ok {
		return nil, fmt.Errorf("no Plan in EXPLAIN output")
	}

	plan, err := ParsePlanJSON(planRaw)
	if err != nil {
		return nil, err
	}

	var execution *ExecutionStats
	if analyze {
		var execTime, planTime float64
		if raw, ok := planObj["Execution Time"]; ok {
			_ = json.Unmarshal(raw, &execTime)
		}
		if raw, ok := planObj["Planning Time"]; ok {
			_ = json.Unmarshal(raw, &planTime)
		}
		execution = &ExecutionStats{
			ExecutionTimeMs: execTime,
			PlanningTimeMs:  planTime,
		}
	}

	warnings := DetectPlanWarnings(plan, snap)

	return &ExplainResult{
		Plan:          *plan,
		TotalCost:     plan.TotalCost,
		EstimatedRows: plan.PlanRows,
		Warnings:      warnings,
		Execution:     execution,
		GucsReplayed:  replayed,
		GucsSkipped:   skipped,
		RawPlanJSON:   planArray[0],
	}, nil
}
